Weird gradle daemon blocking issue

This is a bit of a weird one to explain but i'll try to cover everything as best as possible, simply put as far as i can tell when running portmaster it seems to be blocking me from starting gradle daemons which are being used to clean or compile android projects.
I can get everything working again easily by stopping the portmaster service or by going into developer mode and switching off process detection.
The way i found this issue was totally by accident and actually caused another issue as well.
I had Android Studio running in the background trying to clean + compile and for some reason due to this blocking issue loads of java processes were created but also left to hang in the background.
The result of that was all cpu cores and memory on the system being consumed almost entirely.
After getting things under control i then found that portmaster freaked which resulted in all of my cpu being consumed again.
It would seem portmaster doesn't like it if you have tons of processes created and then hung in the background in quick succession.

Anywho here is a stack trace from trying to run a gradle clean when portmaster is running.
```
./gradlew clean --stacktrace
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/6.5/userguide/gradle_daemon.html.

FAILURE: Build failed with an exception.

* What went wrong:
The first result from the daemon was empty. Most likely the process died immediately after connection.

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.launcher.daemon.client.DaemonInitialConnectException: The first result from the daemon was empty. Most likely the process died immediately after connection.
        at org.gradle.launcher.daemon.client.DaemonClient.executeBuild(DaemonClient.java:177)
        at org.gradle.launcher.daemon.client.SingleUseDaemonClient.execute(SingleUseDaemonClient.java:55)
        at org.gradle.launcher.daemon.client.SingleUseDaemonClient.execute(SingleUseDaemonClient.java:38)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:56)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:212)
        at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:294)
        at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:266)
        at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:82)
        at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:30)
        at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:92)
        at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:38)
        at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:44)
        at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:26)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:41)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:26)
        at org.gradle.launcher.cli.DefaultCommandLineActionFactory$WithLogging.execute(DefaultCommandLineActionFactory.java:259)
        at org.gradle.launcher.Main.doAction(Main.java:35)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:50)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:60)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:31)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:35)
        at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:108)
        at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:63)


* Get more help at https://help.gradle.org
```
After looking through some logs, when the gradle daemon starts i can see an obvious difference.
With portmaster disabled it creates services on 127.0.0.1 but with portmaster running it instead creates services on 0.0.0.0 which seems very odd, but definitely can see straight after that point everything fails.
I can try to provide some actual proper logs if needed, just trying to go through some of them currently.
Here is also some info from the portmaster debug log if you need it.
```
Portmaster
version 0.6.5

commit tags/v0.6.5-0-g6c0df1155e5bc19e8b749875ef984c34eeb8b315
built with go1.15.8 (gc) linux/amd64
  using options main.go
  by user@docker
  on 11.02.2021

Platform: ubuntu 20.10
System: ubuntu linux (debian) 20.10
Kernel: 5.10.16-xanmod1 x86_64
```

--- a/firewall/master.go
+++ b/firewall/master.go
@@ -145,7 +145,7 @@ func checkSelfCommunication(ctx context.Context, conn *network.Connection, pkt p
 				SrcPort:  pktInfo.SrcPort,
 				Dst:      pktInfo.Dst,
 				DstPort:  pktInfo.DstPort,
-			})
+			}, true)
 			if err != nil {
 				log.Tracer(ctx).Warningf("filter: failed to find local peer process PID: %s", err)
 			} else {

--- a/nameserver/takeover.go
+++ b/nameserver/takeover.go
@@ -78,7 +78,7 @@ func takeover(resolverIP net.IP, resolverPort uint16) (int, error) {
 		SrcPort:  0,   // do not record direction
 		Dst:      resolverIP,
 		DstPort:  resolverPort,
-	})
+	}, true)
 	if err != nil {
 		// there may be nothing listening on :53
 		return 0, nil

--- a/network/state/lookup.go
+++ b/network/state/lookup.go
@@ -23,19 +23,20 @@ import (
 //       - switch direction to outbound if outbound packet is seen?
 // - IP: Unidentified Process
 
-// Errors
+// Errors.
 var (
 	ErrConnectionNotFound = errors.New("could not find connection in system state tables")
 	ErrPIDNotFound        = errors.New("could not find pid for socket inode")
 )
 
 var (
-	baseWaitTime  = 3 * time.Millisecond
-	lookupRetries = 7 * 2 // Every retry takes two full passes.
+	baseWaitTime      = 3 * time.Millisecond
+	lookupRetries     = 7 * 2 // Every retry takes two full passes.
+	fastLookupRetries = 2 * 2
 )
 
 // Lookup looks for the given connection in the system state tables and returns the PID of the associated process and whether the connection is inbound.
-func Lookup(pktInfo *packet.Info) (pid int, inbound bool, err error) {
+func Lookup(pktInfo *packet.Info, fast bool) (pid int, inbound bool, err error) {
 	// auto-detect version
 	if pktInfo.Version == 0 {
 		if ip := pktInfo.LocalIP().To4(); ip != nil {
@@ -47,31 +48,31 @@ func Lookup(pktInfo *packet.Info) (pid int, inbound bool, err error) {
 
 	switch {
 	case pktInfo.Version == packet.IPv4 && pktInfo.Protocol == packet.TCP:
-		return tcp4Table.lookup(pktInfo)
+		return tcp4Table.lookup(pktInfo, fast)
 
 	case pktInfo.Version == packet.IPv6 && pktInfo.Protocol == packet.TCP:
-		return tcp6Table.lookup(pktInfo)
+		return tcp6Table.lookup(pktInfo, fast)
 
 	case pktInfo.Version == packet.IPv4 && pktInfo.Protocol == packet.UDP:
-		return udp4Table.lookup(pktInfo)
+		return udp4Table.lookup(pktInfo, fast)
 
 	case pktInfo.Version == packet.IPv6 && pktInfo.Protocol == packet.UDP:
-		return udp6Table.lookup(pktInfo)
+		return udp6Table.lookup(pktInfo, fast)
 
 	default:
 		return socket.UnidentifiedProcessID, false, errors.New("unsupported protocol for finding process")
 	}
 }
 
-func (table *tcpTable) lookup(pktInfo *packet.Info) (
+func (table *tcpTable) lookup(pktInfo *packet.Info, fast bool) (
 	pid int,
 	inbound bool,
 	err error,
 ) {
-	// Search pattern: search, wait, search, refresh, search, wait, search, refresh, ...
+	// Search pattern: search, refresh, search, wait, search, refresh, search, wait, ...
 
 	// Search for the socket until found.
-	for i := 0; i <= lookupRetries; i++ {
+	for i := 1; i <= lookupRetries; i++ {
 		// Check main table for socket.
 		socketInfo, inbound := table.findSocket(pktInfo)
 		if socketInfo == nil && table.dualStack != nil {
@@ -83,6 +84,11 @@ func (table *tcpTable) lookup(pktInfo *packet.Info) (
 		// If there's a match, check we have the PID and return.
 		if socketInfo != nil {
 			return checkPID(socketInfo, inbound)
+		}
+
+		// Search less if we want to be fast.
+		if fast && i < fastLookupRetries {
+			break
 		}
 
 		// every time, except for the last iteration
@@ -134,12 +140,12 @@ func (table *tcpTable) findSocket(pktInfo *packet.Info) (
 	return nil, false
 }
 
-func (table *udpTable) lookup(pktInfo *packet.Info) (
+func (table *udpTable) lookup(pktInfo *packet.Info, fast bool) (
 	pid int,
 	inbound bool,
 	err error,
 ) {
-	// Search pattern: search, wait, search, refresh, search, wait, search, refresh, ...
+	// Search pattern: search, refresh, search, wait, search, refresh, search, wait, ...
 
 	// TODO: Currently broadcast/multicast scopes are not checked, so we might
 	// attribute an incoming broadcast/multicast packet to the wrong process if
@@ -148,7 +154,7 @@ func (table *udpTable) lookup(pktInfo *packet.Info) (
 	isInboundMulticast := pktInfo.Inbound && netutils.ClassifyIP(pktInfo.LocalIP()) == netutils.LocalMulticast
 
 	// Search for the socket until found.
-	for i := 0; i <= lookupRetries; i++ {
+	for i := 1; i <= lookupRetries; i++ {
 		// Check main table for socket.
 		socketInfo := table.findSocket(pktInfo, isInboundMulticast)
 		if socketInfo == nil && table.dualStack != nil {
@@ -171,6 +177,11 @@ func (table *udpTable) lookup(pktInfo *packet.Info) (
 
 			// Check we have the PID and return.
 			return checkPID(socketInfo, connInbound)
+		}
+
+		// Search less if we want to be fast.
+		if fast && i < fastLookupRetries {
+			break
 		}
 
 		// every time, except for the last iteration

--- a/process/find.go
+++ b/process/find.go
@@ -19,9 +19,13 @@ func GetProcessByConnection(ctx context.Context, pktInfo *packet.Info) (process 
 		return GetUnidentifiedProcess(ctx), pktInfo.Inbound, nil
 	}
 
+	// Use fast search for inbound packets, as the listening socket should
+	// already be there for a while now.
+	fastSearch := pktInfo.Inbound
+
 	log.Tracer(ctx).Tracef("process: getting pid from system network state")
 	var pid int
-	pid, connInbound, err = state.Lookup(pktInfo)
+	pid, connInbound, err = state.Lookup(pktInfo, fastSearch)
 	if err != nil {
 		log.Tracer(ctx).Debugf("process: failed to find PID of connection: %s", err)
 		return nil, pktInfo.Inbound, err

